src/range.rs

```diff
@@ -7,354 +7,455 @@
+                .expect("if there is a first element, there must be a last element");
+            (bound_as_ref(start), bound_as_ref(&end.1))
+        })
+    }
+
+    /// Returns true if the this Range contains the specified value.
+    pub fn contains(&self, v: &V) -> bool {
+        for segment in self.segments.iter() {
+            if match segment {
+                (Unbounded, Unbounded) => true,
+                (Unbounded, Included(end)) => v <= end,
+                (Unbounded, Excluded(end)) => v < end,
+                (Included(start), Unbounded) => v >= start,
+                (Included(start), Included(end)) => v >= start && v <= end,
+                (Included(start), Excluded(end)) => v >= start && v < end,
+                (Excluded(start), Unbounded) => v > start,
+                (Excluded(start), Included(end)) => v > start && v <= end,
+                (Excluded(start), Excluded(end)) => v > start && v < end,
+            } {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Construct a simple range from anything that impls [RangeBounds] like `v1..v2`.
+    pub fn from_range_bounds<R, IV>(bounds: R) -> Self
+    where
+        R: RangeBounds<IV>,
+        IV: Clone + Into<V>,
+    {
+        let start = match bounds.start_bound() {
+            Included(v) => Included(v.clone().into()),
+            Excluded(v) => Excluded(v.clone().into()),
+            Unbounded => Unbounded,
+        };
+        let end = match bounds.end_bound() {
+            Included(v) => Included(v.clone().into()),
+            Excluded(v) => Excluded(v.clone().into()),
+            Unbounded => Unbounded,
+        };
+        match (start, end) {
+            (Included(a), Included(b)) if b < a => Self::empty(),
+            (Excluded(a), Excluded(b)) if b < a => Self::empty(),
+            (Included(a), Excluded(b)) if b <= a => Self::empty(),
+            (Excluded(a), Included(b)) if b <= a => Self::empty(),
+            (a, b) => Self {
+                segments: SmallVec::one((a, b)),
+            },
+        }
+    }
+}
 
-    // Union and intersection ##################################################
+/// Implementation of [`Bound::as_ref`] which is currently marked as unstable.
+fn bound_as_ref<V>(bound: &Bound<V>) -> Bound<&V> {
+    match bound {
+        Included(v) => Included(v),
+        Excluded(v) => Excluded(v),
+        Unbounded => Unbounded,
+    }
+}
 
-    /// Compute the union of two sets of versions.
+impl<V: Ord + Clone> Range<V> {
+    /// Computes the union of this `Range` and another.
     pub fn union(&self, other: &Self) -> Self {
-        self.negate().intersection(&other.negate()).negate()
+        self.complement()
+            .intersection(&other.complement())
+            .complement()
     }
 
-    /// Compute the intersection of two sets of versions.
+    /// Computes the intersection of two sets of versions.
     pub fn intersection(&self, other: &Self) -> Self {
```

Comment:
I spent time today trying to write a "more obviously correct" `intersection`, with the same perf. I did not succeed. I did find this method helpful for catching corner cases:

```rust
    fn check_invariants(&self) {
        if cfg!(debug_assertions) {
            for (i, (s, e)) in self.segments.iter().enumerate() {
                if matches!(s, Unbounded) && i != 0 {
                    panic!()
                }
                if matches!(e, Unbounded) && i != (self.segments.len() - 1) {
                    panic!()
                }
            }
            for p in self.segments.as_slice().windows(2) {
                match (&p[0].1, &p[1].0) {
                    (Included(l_end), Included(r_start)) => assert!(l_end < r_start),
                    (Included(l_end), Excluded(r_start)) => assert!(l_end < r_start),
                    (Excluded(l_end), Included(r_start)) => assert!(l_end < r_start),
                    (Excluded(l_end), Excluded(r_start)) => assert!(l_end <= r_start),
                    (_, Unbounded) => panic!(),
                    (Unbounded, _) => panic!(),
                }
            }
            for (s, e) in self.segments.iter() {
                assert!(match (s, e) {
                    (Included(s), Included(e)) => s <= e,
                    (Included(s), Excluded(e)) => s < e,
                    (Excluded(s), Included(e)) => s < e,
                    (Excluded(s), Excluded(e)) => s < e,
                    (Unbounded, _) | (_, Unbounded) => true,
                });
            }
        }
    }
```

Reply:
This is as close as I came:
https://github.com/pubgrub-rs/pubgrub/commit/70b4fdfa0e73ca69787d8e64b2cdae0ce17d178f?diff=split

perf is slightly worse, but I find it more readable. what do you think?

Reply:
I do like that its a lot shorter and easier to follow, my code had a lot of cases. How much worse is the performance?

Reply:
Rerunning the benchmarks now I am not seeing significant differences between our two implementations! I guess I shouldn't try benchmarking at 2 o'clock in the morning.

Reply:
Haha, Ill copy your implementation into the MR. Should I also add the invariant check?

Reply:
Or can I simply merge your changes?

Reply:
I like the check invariants function! Is the first `for` check not included in the second already btw?

Reply:
I cherry-picked your code (crediting you) and added the suggestion about `next_if` by @mpizenberg